File: occonfig-node/__tests__/select.test.ts
import { selectTraceExporter, selectStatsExporter, ExporterType } from "../src/select";

describe("selectTraceExporter", () => {
    type TestCase = {
        source: string,
        type: ExporterType,
        host: string
    }
    const testcases: TestCase[] = [
        { source: "stackdriver://my-project-id", type: ExporterType.STACKDRIVER, host: "my-project-id" },
        { source: "sd://my-project-id", type: ExporterType.STACKDRIVER, host: "my-project-id" },
        { source: "jaeger://localhost:14268", type: ExporterType.JAEGER, host: "http://localhost:14268/api/traces" },
        { source: "jaeger://localhost", type: ExporterType.JAEGER, host: "http://localhost:14268/api/traces" },
        { source: "jaeger", type: ExporterType.JAEGER, host: "http://localhost:14268/api/traces" },
        { source: "zipkin://localhost:9411/api/v2/spans", type: ExporterType.ZIPKIN, host: "http://localhost:9411/api/v2/spans" },
        { source: "zipkin://localhost:9411", type: ExporterType.ZIPKIN, host: "http://localhost:9411/api/v2/spans" },
        { source: "zipkin://localhost", type: ExporterType.ZIPKIN, host: "http://localhost:9411/api/v2/spans" },
        { source: "zipkin://localhost/", type: ExporterType.ZIPKIN, host: "http://localhost:9411/" },
        { source: "zipkin", type: ExporterType.ZIPKIN, host: "http://localhost:9411/api/v2/spans" },
    ];
    for (const testcase of testcases) {
        it(testcase.source, () => {
            const exporter = selectTraceExporter(testcase.source);
            expect(exporter).toBeTruthy();
            if (exporter !== null) {
                expect(exporter.host).toEqual(testcase.host);
                expect(exporter.type).toEqual(testcase.type);
            }
        })
    }
});

describe("selectStatsExporter", () => {
    type TestCase = {
        source: string,
        type: ExporterType,
        host: string
    }
    const testcases: TestCase[] = [
        { source: "stackdriver://my-project-id", type: ExporterType.STACKDRIVER, host: "my-project-id" },
        { source: "sd://my-project-id", type: ExporterType.STACKDRIVER, host: "my-project-id" },
        { source: "prometheus://:8888", type: ExporterType.PROMETHEUS, host: "http://localhost:8888" },
        { source: "p8s://:8888", type: ExporterType.PROMETHEUS, host: "http://localhost:8888" },
    ];
    for (const testcase of testcases) {
        it(testcase.source, () => {
            const exporter = selectStatsExporter(testcase.source);
            expect(exporter).toBeTruthy();
            if (exporter !== null) {
                expect(exporter.host).toEqual(testcase.host);
                expect(exporter.type).toEqual(testcase.type);
            }
        })
    }
});

File: occonfig/select_test.go
package occonfig

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func TestSelectTraceExporterOK(t *testing.T) {
	testcases := []struct {
		Source string
		Type   ExporterType
		Host   string
	}{
		{"stackdriver://my-project-id", STACKDRIVER, "my-project-id"},
		{"sd://my-project-id", STACKDRIVER, "my-project-id"},
		{"datadog://localhost:8125", DATADOG, "localhost:8125"},
		{"datadog://localhost", DATADOG, "localhost:8125"},
		{"datadog", DATADOG, "localhost:8125"},
		{"dd://localhost:8125", DATADOG, "localhost:8125"},
		{"dd://localhost", DATADOG, "localhost:8125"},
		{"dd", DATADOG, "localhost:8125"},
		{"xray", XRAY, ""},
		{"honeycomb", HONEYCOMB, ""},
		{"jaeger://localhost:14268", JAEGER, "http://localhost:14268/api/traces"},
		{"jaeger://localhost", JAEGER, "http://localhost:14268/api/traces"},
		{"jaeger", JAEGER, "http://localhost:14268/api/traces"},
		{"zipkin://localhost:9411/api/v2/spans", ZIPKIN, "http://localhost:9411/api/v2/spans"},
		{"zipkin://localhost:9411", ZIPKIN, "http://localhost:9411/api/v2/spans"},
		{"zipkin://localhost", ZIPKIN, "http://localhost:9411/api/v2/spans"},
		{"zipkin://localhost/", ZIPKIN, "http://localhost:9411/"},
		{"zipkin", ZIPKIN, "http://localhost:9411/api/v2/spans"},
		{"zap", ZAP, ""},
	}
	for _, testcase := range testcases {
		t.Run(testcase.Source, func(t *testing.T) {
			exporter, err := SelectTraceExporter(testcase.Source)
			assert.Nil(t, err)
			if exporter != nil {
				t.Log(testcase.Source, err)
				assert.Equal(t, testcase.Type, exporter.Type)
				assert.Equal(t, testcase.Host, exporter.Host)
			}
		})
	}
}

func TestSelectStatsExporterOK(t *testing.T) {
	testcases := []struct {
		Source string
		Type   ExporterType
		Host   string
	}{
		{"stackdriver://my-project-id", STACKDRIVER, "my-project-id"},
		{"sd://my-project-id", STACKDRIVER, "my-project-id"},
		{"datadog://localhost:8125", DATADOG, "localhost:8125"},
		{"datadog://localhost", DATADOG, "localhost:8125"},
		{"datadog", DATADOG, "localhost:8125"},
		{"dd://localhost:8125", DATADOG, "localhost:8125"},
		{"dd://localhost", DATADOG, "localhost:8125"},
		{"dd", DATADOG, "localhost:8125"},
		{"prometheus://:8888", PROMETHEUS, "http://:8888"},
		{"p8s://:8888", PROMETHEUS, "http://:8888"},
		{"graphite://:2003", GRAPHITE, "localhost:2003"},
		{"graphite", GRAPHITE, "localhost:2003"},
	}
	for _, testcase := range testcases {
		t.Run(testcase.Source, func(t *testing.T) {
			exporter, err := SelectStatsExporter(testcase.Source)
			assert.Nil(t, err)
			if exporter != nil {
				t.Log(testcase.Source, err)
				assert.Equal(t, testcase.Type, exporter.Type)
				assert.Equal(t, testcase.Host, exporter.Host)
			}
		})
	}
}
